Nullable.md

```diff
@@ -0,0 +1,191 @@
+# JULEP Nullable
+
+- **Title:** Nullable Redesign
+- **Authors:** John Myles White <<jmw@johnmyleswhite.com>>
+- **Created:** December 9, 2016
+- **Status:** work in progress
+
+## Abstract
+
+The representation of null-like constructs in Julia is too complicated. We
+currently have all of the following kinds of null-like constructs:
+
+* A ` Void` type with a singleton value `nothing`,
+* A `Nullable` type,
+* An `#undef` quasi-value that reflects the use of null pointers in Julia's
+    implementation.
+
+This Julep describes a proposal to coalesce all of these concepts into one.
+
+The proposal is to replace everything with a new version of the `Nullable`
+type. This new version of `Nullable` will involve both changing the
+implementation of our `Nullable` type and clarifying the semantics of this
+type by laying out a thorough description of how the type would be used in
+practice.
+
+The core goals for this Julep are to:
+
+* Simplify the language by replacing many similar constructs with a single
+    construct.
+* Ensure that all representations of `null` are first-class citizens of the
+    language. In particular, we will resolve the problems with `#undef`, which
+    is a quasi-value that can be present in types, but cannot be inserted into
+    them.
+* Resolve the lifting problem for `Nullable` types by formalizing
+    general-purpose, extensible machinery for lifting.
+* Make it possible to write optimized code for `Nullable` types that does not
+    call into type inference via `return_types` or require the use of
+    `promote_type`.
+
+The optional goals for this Julep are to:
+
+* Replace the `NullableArray` type with `Array{Nullable}`, while maintaining
+    the ability to pass the raw values storage unit via FFI in the way that
+    `NullableArray` currently allows.
+
+The prerequisites for this Julep are the following:
+
+* Optimizations to the Julia compiler to handle `Union{S, T}` as efficiently as
+    our current `Nullable` type handles the union of `S` and `Void`. These
+    compiler changes may include both memory layout changes and optimization
+    of run-time dispatch among K pre-specified types into simple branches.
+
+## Step 1: Redefining the Nullable Type
+
+Assuming that compiler optimizations are put in place for union types, the
+first step for this Julep is to replace the concrete type `Nullable` with an
+optimized union type. The options for replacing `Nullable{T}` are:
+
+* `Union{T, Void}`
+* `Union{Some{T}, Void}`
+* `Union{T, Null{T}}`
+
+The `Union` type would have the same semantics as unions currently have. Their
+implementation in the scalar case would likely match that of a possible
+`TaggedUnion` type, which represents a discriminated union, but is a concrete
+type that is not a parent of `T`, `Void`, etc.
```

Comment:
I'm going to remove this section as I think it adds complexity without many gains.